feat(design): improve sidebar component

xelaint opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/graycoreio/daffodil/blob/develop/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?
1. Sidebar is missing focus trapping. Users currently need to tab through the page to move focus into sidebar.
2. Sidebar is missing custom header and footer components
3. No documentation on sidebar features and usage

Fixes: #277, #2554


## What is the new behavior?
1. Add focus trapping behavior to sidebar so that focus moves to the first focusable element within a sidebar, specifically for `over` and `under` modes.
2. Clean up animation and sidebar visibility so that closed sidebars are not focusable.
3. Create sidebar header that includes a close button that can be hidden, a header title, and custom slotting for additional content.  Add `closeSidebar()` method to header to link closing code to the close button.
4. Create sidebar footer and add body div so that content in body can overflow naturally while allowing footer to stay at bottom of screen.
5. Clean up design land sidebar
6. Extensive documentation on sidebar features and usage

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

xelaint committed:
feat(design): add sidebar header and footer components

xelaint committed:
feat(design): make focus trapping work and clean up animations of sidebar modes

xelaint committed:
docs(design): add sidebar docs

xelaint commented:
This PR is dependent on #2570

damienwebdev commented on libs/design/src/molecules/sidebar/sidebar/sidebar-theme.scss:
```diff
@@ -3,11 +3,9 @@
 
 	.daff-sidebar {
 		background: $base;
-		color: $font-color;
```
Why isn't this `font-color`?

xelaint replied:
reverted

damienwebdev commented on libs/design/src/molecules/sidebar/sidebar/sidebar.component.scss:
```diff
@@ -51,4 +51,12 @@
+.daff-sidebar {
+	&__body {
+		flex-grow: 1;
+    height: 100%;
```
Formatting.

xelaint replied:
This is just github displaying it weirdly here. Formatting is correct in vscode.

damienwebdev commented on libs/design/src/molecules/sidebar/sidebar-header/sidebar-header.component.ts:
```diff
@@ -0,0 +1,36 @@
+  selector: 'daff-sidebar-header',
+  templateUrl: './sidebar-header.component.html',
+  styleUrls: ['./sidebar-header.component.scss'],
+  encapsulation: ViewEncapsulation.None,
```
Why did you need to set `encapsulation: ViewEncapsulation.None,`?

xelaint replied:
In order to style the title. Otherwise, I'd need to use `ng-deep`.

damienwebdev commented on libs/design/sidebar/examples/src/basic-sidebar/basic-sidebar.component.html:
```diff
@@ -1,11 +1,8 @@
-    <nav daff-navbar>
-      <div>Basic Sidebar</div>
-    </nav>
+  <div class="basic-sidebar__content">
```
Why did the content change?

xelaint replied:
This is just the example? I just condensed and cleaned it up.

damienwebdev commented on apps/design-land/src/app/app.component.scss:
```diff
@@ -3,14 +3,27 @@
+
+  &__sidebar {
+    width: 288px;
+    border-right: 1px solid #e8e8e8;
```
We do this border in a few places. Can we embed this in design somehow?

xelaint replied:
No. This design decision should be up to the end user.

damienwebdev commented on libs/design/src/molecules/sidebar/sidebar-viewport/sidebar-viewport.component.ts:
```diff
@@ -133,7 +131,7 @@ export class DaffSidebarViewportComponent implements AfterContentChecked {
     this._animationState = {
       value: getAnimationState(
-        this.sidebars.reduce((acc: boolean, sidebar) => acc || isViewportContentShifted(sidebar.mode, sidebar.open), false),
+        this.sidebars.reduce((acc: boolean, sidebar) => acc || isViewportContentShifted(sidebar.mode, sidebar.open), false), DaffSidebarModeEnum.Over,
```
This needs to be reviewed. I don't remember why we wrote it. It also needs a test.

damienwebdev reviewed (changes requested):
## Documentation
- [x] Indicate the optionality of sidebar header and footer in docs.
- [x] Document class props of sidebar header
- [x] Document usage of Sidebar nav in various contexts.
- [x] Figure out the demo of the sidebar nav in different spots.
- [ ] Read docs a few a times.
- [x] Add documentation about importing `BrowserAnimationsModule` or `NoopAnimationsModule` to use sidebar
- [ ] Add documentation of sidebar with anchor scrolling element within body

## Bugs 
- [x] Right sidebar has a horizontal scroll bar in over/under 
- [x] Anchor scrolling on page doesn't work because height is set on `<daff-sidebar-viewport>`. Solution is to set height on `<daff-sidebar-viewport>` only if sidebar is open and mode is `under` or `over`

## API Considerations
- [x] Remove button from header, add slots instead.
- [ ] ~~Replace title with an ngIf~~

## Tests
- [ ] Test focusing code

## Legibility
- [x] Review the clarity of the updateAnimationState in the sidebar viewport

xelaint committed:
header updates

xelaint committed:
feat(design-land): clean up design-land sidebar and template structure

xelaint committed:
fix horizontal overflow

xelaint commented:
Document example:
- [ ] Sidebar with anchor scrolling

Issue:
Anchor scrolling doesn't work because of height styles in `<daff-sidebar-viewport>`

Possible solution:
Set height on `<daff-sidebar-viewport>` only if sidebar is open and mode is `under` or `over`

xelaint committed:
fix linting

xelaint committed:
fix horizontal scrollbar

xelaint committed:
add doc about required imports

xelaint committed:
set height on viewport only if sidebar is open and mode is over or under

xelaint committed:
update font color

xelaint committed:
allow design-land sidebar to close on esc

xelaint committed:
clean up transformSidebar.done code

xelaint committed:
add tests

damienwebdev committed:
feat(design): correctly trap and destroy focus sidebar state transisitons

damienwebdev committed:
feat(design): adjust animation state for side-fixed

xelaint commented:
- [x] Clean up the sidebar viewport animation legibility
- [x] Add an example with sticky content
- [x] Height shift in examples for over and under sidebars
- [x] Make nav a separate animated block to prevent focus from scrolling page back to top
- [x] Make `over` and `under` sidebars `position: fixed`
- [x] Figure out how to stop body scroll when `over` and `under` sidebars are open without interferring with body height

damienwebdev committed:
code review and bugfixes

damienwebdev committed:
tmp

xelaint committed:
clean up sidebar-viewport template

xelaint committed:
update styling for bounded nav

xelaint committed:
update docs and add sticky content example

damienwebdev committed:
tmp

xelaint commented:
@damienwebdev I've updated `navDesign` to `navPlacement` and added documentation for this. This PR is good to go after you add your portion of the documentation.

In terms of showing the examples for `navPlacement`, the styles we have for a `daff-sidebar-viewport` that is inside of another `daff-sidebar-viewport` is preventing me from adding a working example of `navPlacement="above"` within the sidebar docs.

xelaint committed:
update navDesign property to navPlacement and add docs

xelaint committed:
fix and add tests

damienwebdev committed:
fix: rebase import errors

damienwebdev committed:
feat: adjust class api to use exact enum value

damienwebdev committed:
docs: clarifying docs

xelaint committed:
update example styles

xelaint committed:
update header styles and cadd logo to design-land sidebar viewport
